dpx/src/lib.rs

```diff
@@ -64,6 +64,25 @@ pub(crate) unsafe fn mfree(ptr: *mut libc::c_void) -> *mut libc::c_void {
     std::ptr::null_mut()
 }
 
+use core::mem::MaybeUninit;
+pub trait FromLEByteSlice {
+    fn from_le_byte_slice(b: &[u8]) -> Self;
+}
+impl FromLEByteSlice for u32 {
+    fn from_le_byte_slice(b: &[u8]) -> Self {
+        let mut dst: [u8; 4] = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
This is undefined behaviour. `assume_init`'s unsafe requires that you guarantee you have already written a valid value into it. (I can't think of a scenario where `uninit().assume_init()` will ever not be UB, correct me if I'm wrong.) At the moment, there is junk in there, and it isn't fixed junk, it's compiler-assumes-could-still-change junk. 

The docs recommend a fallible conversion and unwrapping:

https://doc.rust-lang.org/std/primitive.u32.html#method.from_le_bytes

Reply:
Oh, "will ever not be UB" is wrong. If you are claiming to have initialized a `MaybeUninit` which does not require initialization, then it's cool. In general, though, for other types, you want to see:

```rust
let mut x: MaybeUninit<T> = MaybeUninit::uninit();
// The *mut T in the middle is valid and aligned.
unsafe { x.as_mut_ptr().write(get_some_t()); }
// We did indeed initialize it
let x = unsafe { x.assume_init() };
```

Reply:
![изображение](https://user-images.githubusercontent.com/3072754/64961631-9d7a9680-d89e-11e9-813d-27d6057c76a3.png)

Reply:
The std docs recommendation gets the same result as MaybeUninit, without unsafe or UB. Plus, I'm pretty sure all [the call sites are doing the equivalent of `input.split_at` anyway](https://github.com/crlf0710/tectonic/pull/66/files/42137ca3d53179b663a3c8f93543b10b9b6a10b2#diff-d61145e72e33579fa1112ceeb1be0e97R426-R428), so you can refactor that out too by just making the slices into `let mut p = &buf` and using the u32 reader to move their start point at the same time.

https://rust.godbolt.org/z/swMb-R

![Screen Shot 2019-09-17 at 00 37 55](https://user-images.githubusercontent.com/378760/64967311-9c1f8d00-d8e3-11e9-893c-8fddbaee10b5.png)

Reply:
If you change your last line to
`p[0] as u32 + ((p[1] as u32) << 8) + ((p[2] as u32) << 16) + ((p[3] as u32) << 24)`
you'll get the same assembly for both functions

Reply:
Thinking big picture, this kind of binary-format-consuming code is a good candidate for writing a `nom` parser for, instead of doing by hand. I don't mean from_le_byte_slice on its own, but we're talking about a BMP parser, which... we shouldn't have to rewrite, really. And we don't. It's been done.

I'll just drop this here:

https://lib.rs/crates/image

Reply:
I agree. Can you make PR?

Reply:
It'll be a bit of an effort, but yeah later this week I can have a look at replacing ~3kLOC of PNG, BMP, JPEG and JPEG2000 code, and FromLEByteSlice should hopefully be made redundant in the process.

Reply:
> This is undefined behaviour. 
This is not decided yet. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/71). But yeah it's still better to not rely on it anyway.

> yeah later this week I can have a look at replacing ~3kLOC of PNG, BMP, JPEG and JPEG2000 code
Great, I encourage you to create a new crate or expand on existing crate in the Rust ecosystem to do this, leaving minimum code in dpx itself.